### ISSUE TITLE
5 seconds delay on _first_ external request of process

### ISSUE THREAD

**Issue by vrinek** [docker/for-mac#803]
### Expected behavior

External HTTP requests should not experience arbitrary delays.
### Actual behavior

The _first_ external HTTP request of a process (maybe thread I'm not sure) experiences a delay of 5 seconds. Subsequent requests are not delayed.
#### Diagnostics

```
Docker for Mac: version: 1.12.2-beta28 (71c4a00)
OS X: version 10.11.6 (build: 15G1004)
logs: /tmp/6111BB14-101F-4E88-B4A4-16C483A5ED74/20161014-201044.tar.gz
[OK]     docker-cli
[OK]     virtualization kern.hv_support
[OK]     menubar
[OK]     moby-syslog
[OK]     dns
[OK]     disk
[OK]     system
[OK]     app
[OK]     osxfs
[OK]     virtualization VT-X
[OK]     db
[OK]     slirp
[OK]     logs
[OK]     env
[OK]     vmnetd
[OK]     moby-console
[OK]     moby
[OK]     driver.amd64-linux
```
### Steps to reproduce the behavior

**Note:** all of the commands below, when run outside of docker, do not experience the 5 seconds delay.

**Via curl (one process making one request):**

```
docker run buildpack-deps curl 'http://www.google.com'
```

Curl outputs that the request took about 5 seconds.

**Via curl (one process making three requests):**

```
echo 'url = "http://www.google.com"' >> curl.conf
echo 'url = "http://www.google.com"' >> curl.conf
echo 'url = "http://www.google.com"' >> curl.conf
echo 'curl -K curl.conf > /dev/null' > curl.sh
docker run -v (pwd)/curl.sh:/curl.sh -v (pwd)/curl.conf:/curl.conf buildpack-deps sh /curl.sh
```

The first request experiences a delay of 5 seconds while the next two are fast.

**Via curl (multiple processes making one request each):**

```
docker run -it buildpack-deps bash
root@8194bf2a7870:/# curl 'http://www.google.com' > /dev/null
root@8194bf2a7870:/# curl 'http://www.google.com' > /dev/null
root@8194bf2a7870:/# curl 'http://www.google.com' > /dev/null
```

Each request experience a delay of 5 seconds.

**Via ruby (one single-threaded process making two requests):**

```
docker run ruby ruby -e "require 'net/http';\
  t = Time.now; Net::HTTP.get('www.google.com', '/'); puts Time.now - t;\
  t = Time.now; Net::HTTP.get('www.google.com', '/'); puts Time.now - t"
```

This makes 2 requests. The first takes 5 seconds more than the second one.

**Via nodejs (one multi-threaded process making multiple requests):**

Fetch `client.js` from https://gist.github.com/vrinek/99cfdd8d8d48837fd2fd7db6602f2fd3

```
docker run -v (pwd)/client.js:/client.js node node client.js
```

This script makes 10 requests. The first 4 requests (my CPU has 4 cores) experience a 5 second delay while the rest are fine.

**Via apache benchmark:**

```
time docker run --rm jordi/ab ab -n 2 https://www.google.com/
```

`time` gives me about 6.5s but `ab` correctly calculates the request connection time as a few ms.

**Comment by jasonmp85** [comment]
Here I was thinking I was _insane_. What is going on here?

**Comment by jasonmp85** [comment]
(BTW comparing e.g. `curl` to `ping` shows that this issue is triggered by HTTP (TCP? `ssh` shows similar slowdown) requests but not ICMP)

**Comment by vrinek** [comment]
@jasonmp85 Yep, it's definitely bizarre. Kudos for narrowing it down to TCP.

**Comment by ijc25** [comment]
Thanks all for your reports. A 5s delay has the smell of a DNS issue. There have been many improvements to this code between beta28 and beta30 and also improvements to the diagnostics which we gather to aid with cases like these.

Please would you try and reproduce with the latest beta (which is beta30 at the time of writing) and if you can still reproduce upload a fresh diagnostic and tell us the ID.

Thanks,
Ian.

**Comment by rogaha** [comment]
This issue has been inactive for more than 14 days while marked as `status/0-more-info-needed`. It is being closed due to abandonment. Please feel free to re-open with more information about the problem. 

MORE_INFO_EXPIRY_TIMEOUT

**Comment by vrinek** [comment]
@ijc25 sorry for the late reply, I just tried the same under 1.12.3, same behaviour was observed.

```
Docker for Mac: version: 1.12.3 (583d1b8)
OS X: version 10.12.1 (build: 16B2555)
logs: /tmp/6111BB14-101F-4E88-B4A4-16C483A5ED74/20161214-134247.tar.gz
[OK]     vmnetd
[OK]     dns
[OK]     driver.amd64-linux
[OK]     virtualization VT-X
[OK]     app
[OK]     moby
[OK]     system
[OK]     moby-syslog
[OK]     db
[OK]     env
[OK]     virtualization kern.hv_support
[OK]     slirp
[OK]     osxfs
[OK]     moby-console
[OK]     logs
[OK]     docker-cli
[OK]     menubar
[OK]     disk
```

How do I install the beta version?

**Comment by jasonmp85** [comment]
I didn't reproduce with the latest beta because this bug is so bad it pushed me back to the stable channel because I needed to get work done.

Now it's in the stable channel. Great.

**Comment by jasonmp85** [comment]
Just ran diagnostics; ID is `A4F48C72-1628-404A-BF12-BEE394A01B70`. Version is `1.12.5`

**Comment by jasonmp85** [comment]
I'm in the unfortunate position of needing to use Docker again. Going to open another issue for this.

**Comment by jasonmp85** [comment]
Uh, I think this is fixed for me now with 1.13.0…

But I'm working from home today, where I have a working IPv6 stack (oddly absent at my WeWork office). I have a final hunch switching from one to the other on my laptop may somehow be related, but I'll have to follow up next week. For now, Docker 1.13.0 seems to have maybe finally killed this.
